### ISSUE TITLE
question: can renv capture what version of Quarto is installed?

### ISSUE THREAD

**Issue by arthur-shaw** [rstudio/renv#1143]
Imagine that one creates a tool in R--for example, a package or golem app--that creates reports with Quarto.

For all packages involved in creating report content, renv can capture the details in a lock file.

For Quarto, can one snapshot the Quarto version, even though Quarto CLI (unlike rmarkdown) is not a package?

If this is out of scope for renv, what is the best solution for this type of problem? I see that the quarto package has functions that could help in checking that a user's environment is properly provisioned. The `quarto_path()` function returns the path of one's Quarto install. The `quarto_version()` function fetches the version of Quarto installed. Used in combination, these functions could help us determine whether Quarto is installed and, if so, whether it's version is OK for the tool's needs.

**Comment by kevinushey** [comment]
This is a good idea, and something we've considered as well for pandoc. There's some questions on how broadly we want to scope this, though:

- Record the version(s) of required auxiliary software in the lockfile?
- Validate that the version(s) available match that in the lockfile?
- Provide tooling to install, or activate, the requisite versions of these tools?

I think at least some amount of this does belong in `renv`, even if we don't implement everything all at once.

**Comment by kevinushey** [comment]
It's also a bit awkward to consider how conflicts should be resolved in the case of collaborative projects (what if different users collaborating on a project end up with different versions of pandoc / quarto? what should we do when these users call `renv::snapshot()`?)

**Comment by arthur-shaw** [comment]
This does not sound like an easy problem to solve--or even to define.

While I'm no expert in this area, I do see a few use cases that might have different requirements.

### Case 1: set up environment on a virgin system

If one simply wants to set spin up a computational environment on a virgin system, then renv could simply provide a shopping list of dependencies/requirements. Even if renv is unable to fetch/install non-package dependencies, the shopping list is still valuable--especially if renv can warn in the terminal about the things it can't provision. If renv could fully "hydrate" the project in a strictly virgin environment--for example, fetch pandoc and/or Quarto in the correct versions-- that would be better still.

### Case 2: synchronize the systems of collaborators with pre-existing installations of non-package dependencies

Even for a lay person like me, this seems really hard. If one has to deal with an existing environment, then maybe one needs some sort of installation manager like rig that manages not only multiple installations of R but also multiple installations of pandoc and/or Quarto. (Incidentally, it looks like [rig ](https://github.com/r-lib/rig) tries to respect renv R version requirements if my cursory scan of [this](https://github.com/r-lib/rig/blob/main/src/renv.rs) is right.) Alternatively, one might need to spin up a persistent virtual environment (e.g. container) for each project--one that is isolated from the "parent" system and has its own renv-dictated R, package library, pandoc, Quarto, etc.

**Comment by hadley** [comment]
IMO this is currently out of scope for renv, because renv doesn't generally manage system dependencies; it's specifically about R and python packages.

**Comment by arthur-shaw** [comment]
That's a fair point.

However, I wonder if this would be a useful direction for renv to grow. Already, renv provisions system libraries that operate as build tools--or at least that's my layperson's reading of `renv::equip()`. If renv could simply record common system dependencies for R projects, it might take some of the use pain out of compiling the list of system dependencies needed for other tools to fully rebuild a project's environment (e.g., Docker, Podman, etc.). For all users, it might be helpful to have the versions of frequent dependencies--such as pandoc, Quarto, etc--used by a project. For Linux users, it might be helpful additionally to have the list of system libraries needed to get R working (e.g., curl, libssl, etc.) or to use packages in the lockfile (e.g., imagemagick for `{magick}`). In other words, I think it's unreasonable to expect renv to become a drop-in Docker replacement, but perhaps it could do slightly more to document non-package dependencies for R/Python projects and that a Docker/Podman/etc user will need to include in their container composition file somehow.

**Comment by hadley** [comment]
I totally agree that it would be a useful feature, but it's also a huge feature that it's unlikely that we have the development time to implement. And generally, we'd recommend using pre-build binaries (e.g. from https://packagemanager.posit.co/client/) to avoid much of this problem to begin with.

**Comment by kevinushey** [comment]
We don't necessary to take control of installation + management of Quarto entirely, we could just consider:

- Record the version of quarto used in the lockfile,
- Check (and warn?) if the version of quarto differs from what's encoded in the lockfile.

Something similar could be useful for pandoc.

**Comment by hadley** [comment]
Pandoc issue is #615
